File: server/mouse.py
from abc import ABC, abstractmethod


class MouseBase(ABC):
    @abstractmethod
    def press_left(self): ...

    @abstractmethod
    def release_left(self): ...

    @abstractmethod
    def press_right(self): ...

    @abstractmethod
    def release_right(self): ...

    @abstractmethod
    def press_middle(self): ...

    @abstractmethod
    def release_middle(self): ...

    @abstractmethod
    def click_left(self): ...

    @abstractmethod
    def click_right(self): ...

    @abstractmethod
    def click_middle(self): ...

    @abstractmethod
    def move(self, x, y): ...

    @abstractmethod
    def scroll(self, x, y): ...


class Mouse(MouseBase):
    def __init__(self, strategy: MouseBase):
        self._impl = strategy

    def press_left(self):
        self._impl.press_left()

    def release_left(self):
        self._impl.release_left()

    def press_right(self):
        self._impl.press_right()

    def release_right(self):
        self._impl.release_right()

    def press_middle(self):
        self._impl.press_middle()

    def release_middle(self):
        self._impl.release_middle()

    def click_left(self):
        self._impl.click_left()

    def click_right(self):
        self._impl.click_right()

    def click_middle(self):
        self._impl.click_middle()

    def move(self, x, y):
        self._impl.move(x, y)

    def scroll(self, x, y):
        self._impl.scroll(x, y)

File: server/src-tauri/src/mouse.rs
use windows::Win32::UI::Input::KeyboardAndMouse;

pub trait BaseMouse {
    fn press_left();
    fn release_left();
    fn press_right();
    fn release_right();
    fn press_middle();
    fn release_middle();
    fn click_left();
    fn click_right();
    fn click_middle();
    fn move_relative(x: i32, y: i32);
    fn scroll(x: i32, y: i32);
}

pub struct WindowsMouse;

impl BaseMouse for WindowsMouse {
    fn press_left() {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_LEFTDOWN, 0, 0, 0, 0);
        }
    }

    fn release_left() {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_LEFTUP, 0, 0, 0, 0);
        }
    }

    fn press_right() {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_RIGHTDOWN, 0, 0, 0, 0);
        }
    }

    fn release_right() {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_RIGHTUP, 0, 0, 0, 0);
        }
    }

    fn press_middle() {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_MIDDLEDOWN, 0, 0, 0, 0);
        }
    }

    fn release_middle() {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_MIDDLEUP, 0, 0, 0, 0);
        }
    }

    fn click_left() {
        WindowsMouse::press_left();
        WindowsMouse::release_left();
    }

    fn click_right() {
        WindowsMouse::press_right();
        WindowsMouse::release_right();
    }

    fn click_middle() {
        WindowsMouse::press_middle();
        WindowsMouse::release_middle();
    }

    fn move_relative(x: i32, y: i32) {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_MOVE, x, y, 0, 0);
        }
    }

    fn scroll(x: i32, y: i32) {
        unsafe {
            KeyboardAndMouse::mouse_event(KeyboardAndMouse::MOUSEEVENTF_WHEEL, 0, 0, y, x as usize);
        }
    }
}
